### ISSUE TITLE
EventFileWriter can hang on exception in the background thread

### ISSUE THREAD

**Issue by crassirostris** [tensorflow/tensorboard#6167]
### Diagnostics

<details>
<summary>Diagnostics output</summary>

``````
--- check: autoidentify
INFO: diagnose_tensorboard.py version 516a2f9433ba4f9c3a4fdb0f89735870eda054a1

--- check: general
INFO: sys.version_info: sys.version_info(major=3, minor=10, micro=6, releaselevel='final', serial=0)
INFO: os.name: posix
INFO: os.uname(): posix.uname_result(sysname='Linux', nodename='...', release='4.19.128-microsoft-standard', version='#1 SMP Tue Jun 23 12:58:10 UTC 2020', machine='x86_64')
INFO: sys.getwindowsversion(): N/A

--- check: package_management
INFO: has conda-meta: False
INFO: $VIRTUAL_ENV: '/home/mik/Work/tensorboard/tf'

--- check: installed_packages
INFO: installed: tb-nightly==2.12.0a20230123
INFO: installed: tf-nightly==2.12.0.dev20230123
INFO: installed: tf-estimator-nightly==2.12.0.dev2023012321
INFO: installed: tensorboard-data-server==0.6.1

--- check: tensorboard_python_version
INFO: tensorboard.version.VERSION: '2.12.0a20230123'

--- check: tensorflow_python_version
2023-01-24 02:13:42.485983: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2023-01-24 02:13:42.581983: W tensorflow/tsl/platform/default/dso_loader.cc:67] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory
2023-01-24 02:13:42.582026: I tensorflow/tsl/cuda/cudart_stub.cc:28] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
2023-01-24 02:13:42.610952: E tensorflow/tsl/lib/monitoring/collection_registry.cc:81] Cannot register 2 metrics with the same name: /tensorflow/core/bfc_allocator_delay
2023-01-24 02:13:43.168430: W tensorflow/tsl/platform/default/dso_loader.cc:67] Could not load dynamic library 'libnvinfer.so.8'; dlerror: libnvinfer.so.8: cannot open shared object file: No such file or directory
2023-01-24 02:13:43.168504: W tensorflow/tsl/platform/default/dso_loader.cc:67] Could not load dynamic library 'libnvinfer_plugin.so.8'; dlerror: libnvinfer_plugin.so.8: cannot open shared object file: No such file or directory
2023-01-24 02:13:43.168532: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Cannot dlopen some TensorRT libraries. If you would like to use Nvidia GPU with TensorRT, please make sure the missing libraries mentioned above are installed properly.
INFO: tensorflow.__version__: '2.12.0-dev20230123'
INFO: tensorflow.__git_version__: 'v1.12.1-88151-gd5695452f03'

--- check: tensorboard_data_server_version
INFO: data server binary: '/home/mik/Work/tensorboard/tf/lib/python3.10/site-packages/tensorboard_data_server/bin/server'
INFO: data server binary version: b'rustboard 0.6.1'

--- check: tensorboard_binary_path
INFO: which tensorboard: b'/home/mik/Work/tensorboard/tf/bin/tensorboard\n'

--- check: addrinfos
socket.has_ipv6 = True
socket.AF_UNSPEC = <AddressFamily.AF_UNSPEC: 0>
socket.SOCK_STREAM = <SocketKind.SOCK_STREAM: 1>
socket.AI_ADDRCONFIG = <AddressInfo.AI_ADDRCONFIG: 32>
socket.AI_PASSIVE = <AddressInfo.AI_PASSIVE: 1>
Loopback flags: <AddressInfo.AI_ADDRCONFIG: 32>
Loopback infos: [(<AddressFamily.AF_INET6: 10>, <SocketKind.SOCK_STREAM: 1>, 6, '', ('::1', 0, 0, 0)), (<AddressFamily.AF_INET: 2>, <SocketKind.SOCK_STREAM: 1>, 6, '', ('127.0.0.1', 0))]
Wildcard flags: <AddressInfo.AI_PASSIVE: 1>
Wildcard infos: [(<AddressFamily.AF_INET: 2>, <SocketKind.SOCK_STREAM: 1>, 6, '', ('0.0.0.0', 0)), (<AddressFamily.AF_INET6: 10>, <SocketKind.SOCK_STREAM: 1>, 6, '', ('::', 0, 0, 0))]

--- check: readable_fqdn
INFO: socket.getfqdn(): '...'

--- check: stat_tensorboardinfo
INFO: directory: /tmp/.tensorboard-info
INFO: .tensorboard-info directory does not exist

--- check: source_trees_without_genfiles
INFO: tensorboard_roots (2): ['', '/home/mik/Work/tensorboard/tf/lib/python3.10/site-packages']; bad_roots (1): ['']

--- check: full_pip_freeze
INFO: pip freeze --all:
absl-py==1.4.0
appdirs==1.4.4
astunparse==1.6.3
aws-xray-sdk==0.95
black==22.6.0
boto==2.49.0
boto3==1.9.86
botocore==1.12.253
cachetools==5.3.0
certifi==2022.12.7
cffi==1.15.1
charset-normalizer==3.0.1
click==8.1.3
cryptography==39.0.0
distlib==0.3.6
docker==5.0.3
docutils==0.15.2
ecdsa==0.18.0
entrypoints==0.3
filelock==3.9.0
flake8==3.7.8
flatbuffers==23.1.21
fsspec==0.7.4
future==0.18.3
gast==0.4.0
google-auth==2.16.0
google-auth-oauthlib==0.4.6
google-pasta==0.2.0
grpcio==1.51.1
grpcio-testing==1.24.3
h5py==3.8.0
idna==3.4
jax==0.4.1
Jinja2==3.1.2
jmespath==0.10.0
jsondiff==1.1.1
jsonpickle==3.0.1
keras-nightly==2.12.0.dev2023012308
libclang==15.0.6.1
Markdown==3.4.1
MarkupSafe==2.1.2
mccabe==0.6.1
mock==5.0.1
moto==1.3.7
mypy-extensions==0.4.3
numpy==1.23.5
oauthlib==3.2.2
opt-einsum==3.3.0
packaging==23.0
pandas==1.5.3
pathspec==0.10.3
pip==22.3.1
platformdirs==2.6.2
protobuf==3.19.6
pyaml==21.10.1
pyasn1==0.4.8
pyasn1-modules==0.2.8
pycodestyle==2.5.0
pycparser==2.21
pycryptodome==3.16.0
pyflakes==2.1.1
python-dateutil==2.8.2
python-jose==2.0.2
pytz==2022.7.1
PyYAML==6.0
requests==2.28.2
requests-oauthlib==1.3.1
responses==0.22.0
rsa==4.9
s3transfer==0.1.13
scipy==1.10.0
setuptools==59.6.0
six==1.16.0
tb-nightly==2.12.0a20230123
tensorboard-data-server==0.6.1
tensorboard-plugin-wit==1.8.1
tensorflow-io-gcs-filesystem==0.30.0
termcolor==2.2.0
tf-estimator-nightly==2.12.0.dev2023012321
tf-nightly==2.12.0.dev20230123
toml==0.10.2
tomli==2.0.1
types-toml==0.10.8.1
typing_extensions==4.4.0
urllib3==1.25.11
virtualenv==20.0.31
websocket-client==1.4.2
Werkzeug==2.2.2
wheel==0.37.1
wrapt==1.14.1
xmltodict==0.13.0
yamllint==1.17.0

``````

</details>

## Issue description

Right now exception in `_AsyncWriterThread` used in `EventFileWriter` can lead to a deadlock in the thread that's calling `EventFileWriter.add_event`. The sequence of event to cause a deadlock:

1. `EventFileWriter.add_event` is called once in foreground thread, event is added to the queue in `_AsyncWriter`
2. In `_AsyncWriterThread` the event is popped from the queue and record writer starts to write it
3. `EventFileWriter.add_event` is called from the foreground thread enough times to fill the underlying queue in `_AsyncWriter` and the foreground thread then blocks on putting the item to the queue
4. Record writer raises an exception, `_AsyncWriterThread` dies

After this `EventFileWriter.add_event` hangs forever trying to put to the full queue

**Comment by groszewn** [comment]
#6168 should handle this, thanks @crassirostris!
